File: web/lib/leadQualifyRules.ts
import { LeadSource, LeadType } from '@/lib/types';

const DISALLOWED_SOURCES_FOR_QUALIFY = new Set<LeadSource>([
  LeadSource.MANUAL_ENTRY,
  LeadSource.OTHER,
]);

export const STAFF_SELECTABLE_LEAD_SOURCES = Object.values(LeadSource).filter(
  (s) => s !== LeadSource.WEBSITE && !DISALLOWED_SOURCES_FOR_QUALIFY.has(s)
);

export const SELECTABLE_LEAD_TYPES = [
  LeadType.STABLES,
  LeadType.SHEDS,
  LeadType.CABINS,
] as const;

export function leadSourceAllowsQualify(source: LeadSource | string | null | undefined): boolean {
  if (!source) return false;
  return !DISALLOWED_SOURCES_FOR_QUALIFY.has(source as LeadSource);
}

export function leadTypeAllowsQualify(leadType: LeadType | string | null | undefined): boolean {
  if (!leadType) return false;
  return leadType !== LeadType.UNKNOWN;
}

export function leadFieldsAllowQualify(
  source: LeadSource | string | null | undefined,
  leadType: LeadType | string | null | undefined
): boolean {
  return leadSourceAllowsQualify(source) && leadTypeAllowsQualify(leadType);
}

export function qualifyFieldsMessage(
  source: LeadSource | string | null | undefined,
  leadType: LeadType | string | null | undefined
): string {
  const parts: string[] = [];
  if (!leadSourceAllowsQualify(source)) {
    parts.push('select a lead source (not Manual entry or Other)');
  }
  if (!leadTypeAllowsQualify(leadType)) {
    parts.push('select a lead type (Stables, Sheds, or Cabins)');
  }
  return parts.length ? `Before qualifying: ${parts.join('; ')}.` : '';
}

/** Options for a Select including a legacy value not in the selectable list. */
export function leadSourceSelectOptions(current: LeadSource | string | null | undefined): LeadSource[] {
  const base = [...STAFF_SELECTABLE_LEAD_SOURCES];
  if (
    current &&
    !base.includes(current as LeadSource) &&
    Object.values(LeadSource).includes(current as LeadSource)
  ) {
    return [current as LeadSource, ...base];
  }
  return base;
}

export function leadTypeSelectOptions(current: LeadType | string | null | undefined): LeadType[] {
  const base = [...SELECTABLE_LEAD_TYPES];
  if (current === LeadType.UNKNOWN) {
    return [LeadType.UNKNOWN, ...base];
  }
  return base;
}

File: api/app/lead_qualify_rules.py
"""Rules for lead source/type before a lead may become QUALIFIED."""
from typing import Optional

from app.models import LeadSource, LeadType

DISALLOWED_LEAD_SOURCES_FOR_QUALIFY = frozenset({LeadSource.MANUAL_ENTRY, LeadSource.OTHER})

STAFF_SELECTABLE_LEAD_SOURCES: tuple[LeadSource, ...] = tuple(
    s
    for s in LeadSource
    if s not in DISALLOWED_LEAD_SOURCES_FOR_QUALIFY and s != LeadSource.WEBSITE
)

SELECTABLE_LEAD_TYPES: tuple[LeadType, ...] = (
    LeadType.STABLES,
    LeadType.SHEDS,
    LeadType.CABINS,
)


def lead_source_allows_qualify(source: LeadSource | None) -> bool:
    return source not in DISALLOWED_LEAD_SOURCES_FOR_QUALIFY


def lead_type_allows_qualify(lead_type: LeadType | None) -> bool:
    return lead_type is not None and lead_type != LeadType.UNKNOWN


def lead_fields_allow_qualify(
    source: LeadSource | None,
    lead_type: LeadType | None,
) -> bool:
    return lead_source_allows_qualify(source) and lead_type_allows_qualify(lead_type)


def qualify_fields_error(
    source: LeadSource | None,
    lead_type: LeadType | None,
) -> Optional[dict]:
    """Stable HTTP 400 detail when qualify prerequisites are not met."""
    if lead_fields_allow_qualify(source, lead_type):
        return None
    parts: list[str] = []
    if not lead_source_allows_qualify(source):
        parts.append("select a lead source (not Manual entry or Other)")
    if not lead_type_allows_qualify(lead_type):
        parts.append("select a lead type (Stables, Sheds, or Cabins)")
    return {
        "error": "LEAD_FIELDS_REQUIRED_FOR_QUALIFY",
        "message": "Before qualifying: " + "; ".join(parts) + ".",
    }


def staff_may_set_lead_source(source: LeadSource | None) -> bool:
    if source is None:
        return True
    return source in STAFF_SELECTABLE_LEAD_SOURCES


def staff_may_set_lead_type(lead_type: LeadType | None) -> bool:
    if lead_type is None:
        return True
    return lead_type in SELECTABLE_LEAD_TYPES
